### ISSUE TITLE
[concurrency]  dereference failed in data races check

### ISSUE THREAD

**Issue by XLiZHI** [esbmc/esbmc#1324]
`esbmc main4.c --data-races-check`
```
#include <pthread.h>

void* t1(void *arg) {
    int* shared = (int*) arg;
    *shared = 4;                          <-------dereference
}

int main() {
    pthread_t id1;
    int shared_data = 0;

    pthread_create(&id1, NULL, t1, &shared_data);

    return 0;
}
```
When performing a data race check, for a dereferencing operation, we should get the object that the pointer points to, for example, in the code above:

`*shared -----> shared_data` 

We currently handle dereference incorrectly in master：
https://github.com/esbmc/esbmc/blob/c9cd2646987a91ad0b96ab287aee0c990a30a3fa/src/goto-programs/rw_set.cpp#L86-L98

As shown in the above code, after converting `expr` to `expr2tc`, we obtain:

```
symbol
* name : c:main4.c@48@F@t1@shared
* renamelev : Level 0
* level1_num : 0
* level2_num : 0
* thread_num : 0
* node_num : 0
* type : pointer
  * subtype : signedbv
      * width : 32
```
The id of this expr2tc is changed to symbol, which means that the `dereference` cannot continue:
https://github.com/esbmc/esbmc/blob/c9cd2646987a91ad0b96ab287aee0c990a30a3fa/src/pointer-analysis/dereference.cpp#L149-L152

**Comment by XLiZHI** [comment]
I tried some repair as follows:

```
  else if(expr.id() == "dereference")
  {
    assert(expr.operands().size() == 1);
    read(expr.op0(), guard);

    exprt tmp(expr.op0());
    exprt tmp1("dereference", expr.op0().type().subtype());
    tmp1.copy_to_operands(tmp);
    
    expr2tc tmp_expr;
    
    migrate_expr(tmp1, tmp_expr);
```
The new `expr2t` as shown below:
```
dereference
* pointer : symbol
    * name : c:main4.c@48@F@t1@shared
    * renamelev : Level 0
    * level1_num : 0
    * level2_num : 0
    * thread_num : 0
    * node_num : 0
    * type : pointer
      * subtype : signedbv
          * width : 32
* type : signedbv
  * width : 32
```
However, this approach  still got an invalid pointer. We are still unable to get object pointer is pointing to.

https://github.com/esbmc/esbmc/blob/c9cd2646987a91ad0b96ab287aee0c990a30a3fa/src/pointer-analysis/dereference.cpp#L477-L480

**Comment by XLiZHI** [comment]
I have compared the situation when symbolic execution, and found that dereference can be successfully dereference at execution time.

symbolic execution:

```
dereference
* pointer : symbol
    * name : c:main4.c@48@F@t1@shared
    * renamelev : Level 1
    * level1_num : 1
    * level2_num : 0
    * thread_num : 1
    * node_num : 0
    * type : pointer
      * subtype : signedbv
          * width : 32
* type : signedbv
  * width : 32
```

goto program data races check:
```
dereference
* pointer : symbol
    * name : c:main4.c@48@F@t1@shared
    * renamelev : Level 0                                        <---------Different
    * level1_num : 0                                             <---------Different
    * level2_num : 0
    * thread_num : 0                                             <---------Different
    * node_num : 0
    * type : pointer
      * subtype : signedbv
          * width : 32
* type : signedbv
  * width : 32
```
Perhaps this is the reason?

**Comment by lucasccordeiro** [comment]
@XLiZHI: what happens if you use the option`--show-goto-value-sets`? Can you visualize the object that the pointer points to?

**Comment by XLiZHI** [comment]
Hi, @lucasccordeiro ,

I saw those in the result after using option `--show-symex-value-sets`:
```
tmp_c:main4.c@48@F@t1@shared = {  }
c:main4.c@276@F@main@shared_data?1!0 = { # }
```

However, when I use option `--show-goto-value-sets`, only can see one of them:
```
tmp_c:main4.c@48@F@t1@shared = {  }
```

**Comment by lucasccordeiro** [comment]
@fbrausse and @rafaelsamenezes: is there some way to obtain the set of objects that the pointer points to at the goto level?

If not, we must move our data-race check to the symex level :-(

**Comment by rafaelsamenezes** [comment]
Yeah, but its a bit unstable. I have used it at `common_subexpression_removal ` branch.

**Comment by lucasccordeiro** [comment]
Thanks, @rafaelsamenezes.

@XLiZHI: Could you please check Rafael's branch? You might be able to extract something from there.

**Comment by XLiZHI** [comment]
@lucasccordeiro: Thank you! I'll try to check.

Hi @rafaelsamenezes,  can I confirm that the part you mentioned is here?

https://github.com/esbmc/esbmc/blob/eaa030538065200702fa676ce4c7b829d0231d95/src/goto-programs/abstract-interpretation/common_subexpression_elimination.cpp#L212-L213

**Comment by XLiZHI** [comment]
Hi @rafaelsamenezes, Could you please explain why you don't use the method from `goto_program_dereference.cpp` in your branch? Aren't these methods designed to get the object pointed to by a pointer at the goto level?

**Comment by rafaelsamenezes** [comment]
> Hi @rafaelsamenezes, Could you please explain why you don't use the method from goto_program_dereference.cpp in your branch? Aren't these methods designed to get the object pointed to by a pointer at the goto level?

Mainly because I was not familiar with it. That does seems a better way.

**Comment by XLiZHI** [comment]
Hi,

After talking to Rafael, I did some tweaking and testing. I think I found the root cause of the problem as follows:

We were able to obtain the set of objects that the pointer points to at the GOTO level, the existing method worked fine after I 
made a few changes, but there was still a serious problem.

`pthread_create(&id1, NULL, t1, &shared_data);`: We cannot handle `pthread_create` in the goto program, so the function called through it cannot be properly dereference. 

For example, one case we can handle is to call a function directly:

```
#include <pthread.h>

void* t1(void *arg) {
    int* shared = (int*) arg;
    *shared = 4;                          <-------dereference
}

int main() {
    int shared_data = 0;

    t1(&shared_data);                 <-------call a function directly

    // pthread_create(&id1, NULL, t1, &shared_data); <----- This function call cannot be processed.

    return 0;
}
```

To sum up, I think we can probably solve the problem as follows:

1. Perhaps `pthread_create` should be adjusted to optimize this dereference.

2. Move data-races-check to symbolic execution.(Not sure if it's feasible)

@rafaelsamenezes, Can I ask if we can adjust `pthread_create` as a function call?

**Comment by XLiZHI** [comment]
We can see the difference between the two using goto functions:

```
        FUNCTION_CALL:  pthread_create(&id1, 0, &t1, (void *)(&shared_data))
       
        FUNCTION_CALL:  t1((void *)(&shared_data))
```

**Comment by XLiZHI** [comment]
close via #1460
